File: instrumentation/ktor-utils-2.0/src/main/java/io/ktor/util/pipeline/Pipeline_Instrumentation.java
/*
 *
 *  * Copyright 2026 New Relic Corporation. All rights reserved.
 *  * SPDX-License-Identifier: Apache-2.0
 *
 */

package io.ktor.util.pipeline;

import com.newrelic.api.agent.NewRelic;
import com.newrelic.api.agent.Token;
import com.newrelic.api.agent.Trace;
import com.newrelic.api.agent.weaver.MatchType;
import com.newrelic.api.agent.weaver.NewField;
import com.newrelic.api.agent.weaver.Weave;
import com.newrelic.api.agent.weaver.Weaver;
import com.newrelic.instrumentation.labs.ktor.utils.PipelineUtils;
import kotlin.coroutines.Continuation;

@Weave(type = MatchType.BaseClass, originalName = "io.ktor.util.pipeline.Pipeline")
public class Pipeline_Instrumentation<TSubject, TContext> {

    @NewField
    private Token token;

    public Pipeline_Instrumentation(PipelinePhase... phases) {
        if (token != null) {
            return;
        }
        String simpleName = getClass().getSimpleName();
        if (PipelineUtils.tracePipeline(simpleName)) {
            Token t = NewRelic.getAgent().getTransaction().getToken();
            if (t != null && t.isActive()) {
                token = t;
            } else if (t != null) {
                t.expire();
            }
        }
    }

    @Trace(async = true)
    public Object execute(TContext context, TSubject subject, Continuation<? super TSubject> continuation) {
        Token localToken = token;
        token = null;
        if (localToken != null) {
            localToken.linkAndExpire();
        }
        return Weaver.callOriginal();
    }



}

File: instrumentation/ktor-utils-1.4/src/main/kotlin/io/ktor/util/pipeline/Pipeline_Instrumentation.kt
package io.ktor.util.pipeline

import com.newrelic.api.agent.NewRelic
import com.newrelic.api.agent.Trace
import com.newrelic.api.agent.weaver.MatchType
import com.newrelic.api.agent.weaver.Weave
import com.newrelic.api.agent.weaver.Weaver

@Weave(type = MatchType.BaseClass, originalName = "io.ktor.util.pipeline.Pipeline")
open class Pipeline_Instrumentation<TSubject : Any, TContext : Any> {

    @Trace
    suspend fun execute(context : TContext, subject : TSubject) : TSubject  {
        NewRelic.getAgent().tracedMethod.setMetricName("Custom","Ktor-Utils","Pipeline",this.javaClass.name,"execute")
        return Weaver.callOriginal()
    }

    @Trace
    fun intercept(phase: PipelinePhase, block: PipelineInterceptor<TSubject, TContext>) {
        val phaseName = phase.name
        NewRelic.getAgent().tracedMethod.setMetricName("Custom","Ktor-Utils","Pipeline",this.javaClass.simpleName,"intercept",phaseName)
        return Weaver.callOriginal()
    }
}

typealias PipelineInterceptor<TSubject, TContext> = suspend PipelineContext<TSubject, TContext>.(TSubject) -> Unit